File: app/src/main/java/com/freddygenicho/sample/mpesa/MainActivity.java
package com.freddygenicho.sample.mpesa;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.widget.EditText;
import android.widget.Toast;

import com.freddygenicho.mpesa.stkpush.Mode;
import com.freddygenicho.mpesa.stkpush.api.response.STKPushResponse;
import com.freddygenicho.mpesa.stkpush.interfaces.STKListener;
import com.freddygenicho.mpesa.stkpush.interfaces.TokenListener;
import com.freddygenicho.mpesa.stkpush.model.Mpesa;
import com.freddygenicho.mpesa.stkpush.model.STKPush;
import com.freddygenicho.mpesa.stkpush.model.Token;
import com.freddygenicho.mpesa.stkpush.model.Transaction;

import java.io.UnsupportedEncodingException;

import cn.pedant.SweetAlert.SweetAlertDialog;

public class MainActivity extends AppCompatActivity implements TokenListener {

    public static final String TAG = MainActivity.class.getSimpleName();

    private EditText phoneET, amountET;
    private SweetAlertDialog sweetAlertDialog;
    private Mpesa mpesa;

    private String phone_number;
    private String amount;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        phoneET = findViewById(R.id.phoneET);
        amountET = findViewById(R.id.amountET);

        mpesa = new Mpesa(Config.CONSUMER_KEY, Config.CONSUMER_SECRET, Mode.SANDBOX);

        sweetAlertDialog = new SweetAlertDialog(this, SweetAlertDialog.PROGRESS_TYPE);
        sweetAlertDialog.setTitleText("Connecting to Safaricom");
        sweetAlertDialog.setContentText("Please wait...");
        sweetAlertDialog.setCancelable(false);
    }

    public void startMpesa(View view) {

        phone_number = phoneET.getText().toString();
        amount = amountET.getText().toString();

        if (phone_number.isEmpty()) {
            Toast.makeText(MainActivity.this, "Phone Number is required", Toast.LENGTH_SHORT).show();
            return;
        }

        if (amount.isEmpty()) {
            Toast.makeText(MainActivity.this, "Amount is required", Toast.LENGTH_SHORT).show();
            return;
        }

        if (!phone_number.isEmpty() && !amount.isEmpty()) {
            try {
                sweetAlertDialog.show();
                mpesa.getToken(this);
            } catch (UnsupportedEncodingException e) {
                Log.e(TAG, "UnsupportedEncodingException: " + e.getLocalizedMessage());
            }
        } else {
            Toast.makeText(MainActivity.this, "Please make sure that phone number and amount is not empty ", Toast.LENGTH_SHORT).show();
        }
    }

    @Override
    public void onTokenSuccess(Token token) {
        STKPush stkPush = new STKPush();
        stkPush.setBusinessShortCode(Config.BUSINESS_SHORT_CODE);
        stkPush.setPassword(STKPush.getPassword(Config.BUSINESS_SHORT_CODE, Config.PASSKEY, STKPush.getTimestamp()));
        stkPush.setTimestamp(STKPush.getTimestamp());
        stkPush.setTransactionType(Transaction.CUSTOMER_PAY_BILL_ONLINE);
        stkPush.setAmount(amount);
        stkPush.setPartyA(STKPush.sanitizePhoneNumber(phone_number));
        stkPush.setPartyB(Config.PARTYB);
        stkPush.setPhoneNumber(STKPush.sanitizePhoneNumber(phone_number));
        stkPush.setCallBackURL(Config.CALLBACKURL);
        stkPush.setAccountReference("test");
        stkPush.setTransactionDesc("some description");

        mpesa.startStkPush(token, stkPush, new STKListener() {
            @Override
            public void onResponse(STKPushResponse stkPushResponse) {
                Log.e(TAG, "onResponse: " + stkPushResponse.toJson(stkPushResponse));
                String message = "Please enter your pin to complete transaction";
                sweetAlertDialog.changeAlertType(SweetAlertDialog.SUCCESS_TYPE);
                sweetAlertDialog.setTitleText("Transaction started");
                sweetAlertDialog.setContentText(message);
            }

            @Override
            public void onError(Throwable throwable) {
                Log.e(TAG, "stk onError: " + throwable.getMessage());
                sweetAlertDialog.changeAlertType(SweetAlertDialog.ERROR_TYPE);
                sweetAlertDialog.setTitleText("Error");
                sweetAlertDialog.setContentText(throwable.getMessage());
            }
        });
    }

    @Override
    public void OnTokenError(Throwable throwable) {
        Log.e(TAG, "mpesa Error: " + throwable.getMessage());
        sweetAlertDialog.changeAlertType(SweetAlertDialog.ERROR_TYPE);
        sweetAlertDialog.setTitleText("Error");
        sweetAlertDialog.setContentText(throwable.getMessage());
    }
}

File: app/src/main/java/com/freddygenicho/sample/mpesa/MainActivity.kt
package com.freddygenicho.sample.mpesa

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import android.util.Log
import android.view.View
import android.widget.EditText
import android.widget.Toast
import cn.pedant.SweetAlert.SweetAlertDialog
import com.freddygenicho.mpesa.stkpush.Mode
import com.freddygenicho.mpesa.stkpush.api.response.STKPushResponse
import com.freddygenicho.mpesa.stkpush.interfaces.STKListener
import com.freddygenicho.mpesa.stkpush.interfaces.TokenListener
import com.freddygenicho.mpesa.stkpush.model.Mpesa
import com.freddygenicho.mpesa.stkpush.model.STKPush
import com.freddygenicho.mpesa.stkpush.model.Token
import com.freddygenicho.mpesa.stkpush.model.Transaction
import java.io.UnsupportedEncodingException

class MainActivity : AppCompatActivity(), TokenListener {
    private var phoneET: EditText? = null
    private var amountET: EditText? = null
    private var sweetAlertDialog: SweetAlertDialog? = null
    private var mpesa: Mpesa? = null
    private var phoneNumber: String? = null
    private var amount: String? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        phoneET = findViewById(R.id.phoneET)
        amountET = findViewById(R.id.amountET)
        mpesa = Mpesa(Config.CONSUMER_KEY, Config.CONSUMER_SECRET, Mode.SANDBOX)

    }

    fun startMpesa(view: View?) {
        phoneNumber = phoneET!!.text.toString()
        amount = amountET!!.text.toString()
        if (phoneNumber!!.isEmpty()) {
            Toast.makeText(this@MainActivity, "Phone Number is required", Toast.LENGTH_SHORT).show()
            return
        }
        if (amount!!.isEmpty()) {
            Toast.makeText(this@MainActivity, "Amount is required", Toast.LENGTH_SHORT).show()
            return
        }
        if (phoneNumber!!.isNotEmpty() && amount!!.isNotEmpty()) {
            try {
                mpesa!!.getToken(this)
            } catch (e: UnsupportedEncodingException) {
                Log.e(TAG, "UnsupportedEncodingException: " + e.localizedMessage)
            }
        } else {
            Toast.makeText(
                this@MainActivity,
                "Please make sure that phone number and amount is not empty ",
                Toast.LENGTH_SHORT
            ).show()
        }
    }

    override fun onTokenSuccess(token: Token?) {
        sweetAlertDialog = SweetAlertDialog(this, SweetAlertDialog.PROGRESS_TYPE)
        sweetAlertDialog!!.titleText = "Connecting to Mpesa"
        sweetAlertDialog!!.contentText = "Please wait..."
        sweetAlertDialog!!.setCancelable(false)
        sweetAlertDialog!!.show()

        val stkPush = STKPush()
        stkPush.businessShortCode = Config.BUSINESS_SHORT_CODE
        stkPush.password =
            STKPush.getPassword(Config.BUSINESS_SHORT_CODE, Config.PASSKEY, STKPush.getTimestamp())
        stkPush.setTimestamp(STKPush.getTimestamp())
        stkPush.transactionType = Transaction.CUSTOMER_PAY_BILL_ONLINE
        stkPush.amount = amount
        stkPush.partyA = STKPush.sanitizePhoneNumber(phoneNumber!!)
        stkPush.partyB = Config.PARTYB
        stkPush.phoneNumber = STKPush.sanitizePhoneNumber(phoneNumber!!)
        stkPush.callBackURL = Config.CALLBACKURL
        stkPush.accountReference = "test"
        stkPush.transactionDesc = "some description"
        mpesa!!.startStkPush(token, stkPush, object : STKListener {
            override fun onResponse(stkPushResponse: STKPushResponse?) {
                Log.e(TAG, "onResponse: " + stkPushResponse?.toJson(stkPushResponse))
                val message = "Please enter your pin to complete transaction"
                sweetAlertDialog!!.changeAlertType(SweetAlertDialog.SUCCESS_TYPE)
                sweetAlertDialog!!.titleText = "Transaction started"
                sweetAlertDialog!!.contentText = message
            }

            override fun onError(throwable: Throwable?) {
                Log.e(TAG, "stk onError: " + throwable?.message)
                sweetAlertDialog!!.changeAlertType(SweetAlertDialog.ERROR_TYPE)
                sweetAlertDialog!!.titleText = "Error"
                sweetAlertDialog!!.contentText = throwable?.message
            }
        })
    }

    override fun onTokenError(throwable: Throwable?) {
        Log.e(TAG, "mpesa Error: " + throwable!!.message)
        sweetAlertDialog!!.changeAlertType(SweetAlertDialog.ERROR_TYPE)
        sweetAlertDialog!!.titleText = "Error"
        sweetAlertDialog!!.contentText = throwable.message
    }

    companion object {
        val TAG: String = MainActivity::class.java.simpleName
    }
}